### PR TITLE
fix when curl set time less than one second causes libcurl to timeout…

### PR DESCRIPTION
**Pull request by tom916**
http://php.net/manual/en/function.curl-setopt.php

>If you want cURL to timeout in less than one second, you can use CURLOPT_TIMEOUT_MS, although >there is a bug/"feature"  on "Unix-like systems" that causes libcurl to timeout immediately if the value is >< 1000 ms with the error "cURL Error (28): Timeout was reached".  The explanation for this behavior is:

### CODE REVIEW

File: src/Handler/CurlFactory.php [hunk 1]
```diff
@@ -370,8 +370,9 @@ private function applyHandlerOptions(EasyHandle $easy, array &$conf)
             $conf[CURLOPT_FILE] = fopen('php://temp', 'w+');
             $easy->sink = Psr7\stream_for($conf[CURLOPT_FILE]);
         }
-
+        $lessThan1s = false;
```
Comment (mtdowling) [inline src/Handler/CurlFactory.php]: can we call this, `$timeoutRequiresNoSignal` or something more specific? Can you also add a comment that describes why we are doing this and link to the original comment at http://php.net/manual/en/function.curl-setopt.php#104597

File: src/Handler/CurlFactory.php [hunk 2]
```diff
@@ -385,8 +386,12 @@ private function applyHandlerOptions(EasyHandle $easy, array &$conf)
         }
 
         if (isset($options['connect_timeout'])) {
+            $options['connect_timeout'] < 1 && $lessThan1s = true;
             $conf[CURLOPT_CONNECTTIMEOUT_MS] = $options['connect_timeout'] * 1000;
         }
+        if ($lessThan1s) {
+            $conf[CURLOPT_NOSIGNAL] = 1;
```
Comment (mtdowling) [inline src/Handler/CurlFactory.php]: This should be `true` and not `1`.

**Comment by sagikazarmark** [comment]
This seems to be  duplication of #1585

File: src/Handler/CurlFactory.php (hunk 1 above)
Comment (tom916) [inline src/Handler/CurlFactory.php]: @mtdowling  I mean in the url ,there are very detailed explanation

**Comment by tom916** [comment]
@mtdowling 

I add the test case for old version

```
<?php
namespace GuzzleHttp\Tests;
use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
class MillSecondTest extends \PHPUnit_Framework_TestCase
{

    public function testSetTimeoutSmallThanOneSecond()
    {
        $client = new Client();
        $requestOptions = [
                'connect_timeout' => 0.1,
                'timeout' => 0.1
        ];

        try {
            $request = new Request('GET', "http://foo.com");
            $client->send($request, $requestOptions);
        } catch (\Exception $e) {
            $this->assertEquals(true, preg_match("#resolve via SIGALRM method#", $e->getMessage()));
        }
    }
}

```

File: src/Handler/CurlFactory.php [hunk 3]
```diff
@@ -370,8 +370,9 @@ private function applyHandlerOptions(EasyHandle $easy, array &$conf)
             $conf[CURLOPT_FILE] = fopen('php://temp', 'w+');
             $easy->sink = Psr7\stream_for($conf[CURLOPT_FILE]);
         }
-
+        $timeoutRequiresNoSignal = false;
         if (isset($options['timeout'])) {
+            $options['timeout'] < 1 && $timeoutRequiresNoSignal = true;
```
Comment (sagikazarmark) [inline src/Handler/CurlFactory.php]: How about doing `$timeoutRequiresNoSignal |= $options['timeout'] < 1` instead?

File: src/Handler/CurlFactory.php [hunk 4]
```diff
@@ -385,9 +386,14 @@ private function applyHandlerOptions(EasyHandle $easy, array &$conf)
         }
 
         if (isset($options['connect_timeout'])) {
+            $options['connect_timeout'] < 1 && $timeoutRequiresNoSignal = true;
```
Comment (sagikazarmark) [inline src/Handler/CurlFactory.php]: Same here

File: src/Handler/CurlFactory.php (hunk 3 above)
Comment (tom916) [inline src/Handler/CurlFactory.php]: @sagikazarmark 
 be better 👍

**Comment by sagikazarmark** [comment]
I am actually in favor of this instead of #1585.

**Comment by sagikazarmark** [comment]
For the record: here is the full explanation: http://php.net/manual/en/function.curl-setopt.php#104597

**Comment by sagikazarmark** [comment]
@mtdowling this looks good to me. Are you satisfied with your requested changes?

**Comment by sagikazarmark** [comment]
Ping @mtdowling

**Comment by mtdowling** [comment]
:shipit:

**Comment by sagikazarmark** [comment]
Thanks @tom916

**Pull request merged**